### DISCUSSION TITLE
Remember time of day a habit was fulfilled and graph it

### DISCUSSION THREAD

**Discussion by LuccoJ** [iSoron/uhabits#129] [Feature Requests]
Sometimes you only tick habits towards the end of your day, or when the alarm reminds you, but sometimes you want to tick them immediately after fulfilling them. If you do the latter, it can be interesting to know when, during the day, you tend to fulfill your habits.

For UI simplicity, I wouldn't make it possible (or at least default) to manually specify the time, but I would simply make the application remember the time a habit was ticked. Then, in the statistics view, there could be a "Time of day" view along with "Weekly" etc, with the hours when the habit tends to occur the most having peaks in the graph.

**Comment by iSoron** [comment]
Thank you for the suggestion, and sorry about the delay to reply. I am still considering it. In any case, I think the app should only display some data if it also allows you to modify it, so this feature would require changing the UI somehow.

**Reply by boxallar** [reply]
I see that the time stamps are always rounded off. Presumably because for most people whether they did a task or not on _a given day_ is what is important. 
eg. I get this `1711497600000` time stamp for a task completed on 27th March at 08:30. The five 0s at the end show that the stamp has been rounded off to the nearest day. This `1711528251000` would be the stamp rounded to the nearest second.
For those who want to know when something was marked as done, could the APP record unrounded time stamps?
When the app wants to know what day something was marked as done, it can presumably use unrounded stamps just as well.
For people who want the extra level of detail, they can export the data and play with it as they like.
Being able to view the extra information in the app could remain a project for the future.

I have tried to be clear here, but it was not easy, please let me know if you need any clarification.

**Comment by LuccoJ** [comment]
How about a long tap anyway? In the following way:
- if short tapping is being used, then long tapping will show a clock and you can change the time
- if long tapping is being used, then of course it will cause the habit to be ticked, but then if it's long tapped again, instead of being unticked immediately, a clock appears, including a button to "remove" (untick)

**Comment by evegeris** [comment]
How about a feature you can toggle under Settings like long-tap vs, short tap that will prompt the user to enter more information when they check off a habit? In this window, there could be optional fields for duration, record time of day, user notes, etc. If this information is stored in the database, a chart could later be added to the Habit information Activity for tracking time of day. A message box could also be brought up when a given day is clicked in the history chart that would display any entered information for that day.

**Comment by boxallar** [comment]
I have a workaround that may work for some people. 
I make a task with an arbitrary target. When entering the result I first type the time eg. for 5:30 pm I write `1730` and then I can add eg. 0 for no or 1 for yes. So if I did something at half past five I enter `17301`.
This can be extracted and combined with the date time stamp with a bit of fiddling. You could also put a decimal point after the time and/or record a numerical value with a consistent number of digits instead of a 1 or 0.

**Comment by herrbrixter** [comment]
I want to remember the time I updated a habit, yes/no and measurable habits, but I don't need to see a graph. That's too complex.

Related to #1313
